applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -246,43 +302,112 @@ void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveV
+    KRATOS_ERROR_IF( std::abs( Factor ) != 1.0 ) << "MassConservationPocess: Given value of argument 'factor' is not plausible. Consider '1' or '-1'." << std::endl;
+
+    #pragma omp parallel for reduction(+: outflow)
     for (int i_elem = 0; i_elem < static_cast<int>(mrModelPart.NumberOfElements()); ++i_elem){
         // iteration over all elements
         const auto it_elem = mrModelPart.ElementsBegin() + i_elem;
 
         Matrix shape_functions;
         GeometryType::ShapeFunctionsGradientsType shape_derivatives;
 
-        const auto rGeom = it_elem->GetGeometry();
+        const auto r_geom = it_elem->GetGeometry();
         unsigned int pt_count_pos = 0;
         unsigned int pt_count_neg = 0;
+        const bool geom_is_cut = IsGeometryCut(r_geom, pt_count_neg, pt_count_pos);
+        if (geom_is_cut){
+            // element is cut by the surface (splitting)
+            Kratos::unique_ptr<ModifiedShapeFunctions> p_modified_sh_func = nullptr;
+            Vector w_gauss_interface(3, 0.0);
 
-        // instead of using data.isCut()
-        for (unsigned int pt = 0; pt < rGeom.Points().size(); pt++){
-            if ( rGeom[pt].FastGetSolutionStepValue(DISTANCE) > 0.0 ){
-                pt_count_pos++;
-            } else {
-                pt_count_neg++;
+            Vector Distance( r_geom.PointsNumber(), 0.0 );
+            for (unsigned int i = 0; i < r_geom.PointsNumber(); i++){
+                // Control mechanism to avoid 0.0 ( is necessary because "distance_modification" possibly not yet executed )
+                if ( r_geom[i].FastGetSolutionStepValue(DISTANCE) == 0.0 ){
+                    it_elem->GetGeometry().GetPoint(i).FastGetSolutionStepValue(DISTANCE) = -1.0e-7;
+                }
+                Distance[i] = r_geom[i].FastGetSolutionStepValue(DISTANCE);
             }
-        }
 
-        if ( pt_count_pos == rGeom.PointsNumber() || pt_count_neg == rGeom.PointsNumber() ){
-            // all nodes are on one side and the element is not split
-            continue;
+            if ( r_geom.PointsNumber() == 3 ){ p_modified_sh_func = Kratos::make_unique<Triangle2D3ModifiedShapeFunctions>(it_elem->pGetGeometry(), Distance); }
+            else if ( r_geom.PointsNumber() == 4 ){ p_modified_sh_func = Kratos::make_unique<Tetrahedra3D4ModifiedShapeFunctions>(it_elem->pGetGeometry(), Distance); }
+            else { KRATOS_ERROR << "The utility can not be applied on this kind of element" << std::endl; }
+
+            // Concerning their area, the positive and negative side of the interface are equal
+            p_modified_sh_func->ComputeInterfacePositiveSideShapeFunctionsAndGradientsValues(
+                    shape_functions,                    // N
+                    shape_derivatives,                  // DN
+                    w_gauss_interface,                  // includes the weights of the GAUSS points (!!!)
+                    GeometryData::GI_GAUSS_2);          // second order Gauss integration
```

Comment:
This is just a comment. I've observed that there are a lot of cases in which the shape function derivatives are not required. As this operation is performed lots of times, I think we (possibly I) should create a method that only computes the shape function values and weights.

Reply:
Yeeap, that'd be great